### PR TITLE
Topic Classification M1 labeling and training

### PR DESCRIPTION
**Pull request by bdittes**
fixes #

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

### CODE REVIEW

**Comment by gabriel-piles** [comment]
Good job on the topic Classification Implementation. Many functionalities have been implemented to achieve that.

It seems that this first version does not cover mobile, but probably we should fix it before launching it for everyone.

mobile
Screenshot 2020-03-20 at 10 18 07
Screenshot 2020-03-20 at 10 21 47
Screenshot 2020-03-20 at 10 32 04

Using the e2e database, two functionalities were not working well. 

The first one is in the quick label, the save document when no autosave is not working


The second functionality not working well, in the same section the number of documents in each label are not well calculated

**Comment by gabriel-piles** [comment]
Using the e2e database, two functionalities were not working well. 

The first one is in the quick label, the save document when no autosave is not working


The second functionality not working well, in the same section the number of documents in each label are not well calculated

**Comment by bdittes** [comment]
Thanks for the first look!

Re: mobile - I must admit I have 0 expertise there. All mocks I got from Alberto were for desktop and we launched M0 in its current state, so I'd like to declare this out of scope for now. I have made some CSS changes, though, Settings and QuickLabel look good now. OneUpReview does not.

Re: 'any' negative values: I cannot reproduce, maybe it went away after the last merge?

Re: 'save' - fixed.

**Comment by gabriel-piles** [comment]
Some observations:

* **A feature toggle is needed**: the topic classification should be turned off for the instances that we don't feel prepared for it. The actual implementation of features toggles can be used in order to achieve this.

* **When the feature toggle is off, the mobile layout should be displayed well**: in the current implementation, there are some issues with the mobile layout even for some not related topic classification windows. 

* In the quick labeling window:
 1. Why are entities without the target label shown?
 2. Is it a good idea to show a text or a title to indicate that this view is not the regular library?

**Comment by gabriel-piles** [comment]
An error occurs when review and publish an entity when reviewing the suggestions for Super Powers in the e2e database

output

**Comment by bdittes** [comment]
Many updates:
 - mobile CSS with 'Select All' and long thesaurus names should work better
 - any aggregation and save-and-next bugs fixed
 - feature toggle added (also to e2e-fixtures)

> Why are entities without the target label shown?

I found it hard to filter them out...

> Is it a good idea to show a text or a title to indicate that this view is not the regular library?

Yes, I'm planning to add a 'header' band and move the back button there.

File: app/api/entities/entities.js [hunk 1]
```diff
@@ -259,6 +259,30 @@ function sanitize(doc, template) {
   return Object.assign(doc, { metadata });
 }
 
+function applyDiffMetadata(metadata, diffMetadata) {
```
Comment (gabriel-piles) [inline app/api/entities/entities.js]: I would prefer a more clear name for this function

Comment (bdittes) [reply app/api/entities/entities.js]: updateMetadataWithDiff?

File: app/api/tasks/routes.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+export default (app: Application) => {
+  app.get(
+    tasksPrefix,
+    validation.validateRequest(
+      Joi.object()
+        .keys({ name: Joi.string() })
+        .required(),
+      'query'
+    ),
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/api/tasks/routes.ts]: done and moved to supertest.

File: app/api/tasks/specs/routes.spec.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/** @format */
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/api/tasks/specs/routes.spec.ts]: removed in all touched files.

File: app/react/Library/components/specs/QuickLabelHeader.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  });
+
+  const render = () => {
+    component = shallow(<QuickLabelHeaderBase {...props} />);
+  };
+
+  it('should render with thesaurus', () => {
+    render();
+    expect(component).toMatchSnapshot();
+  });
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/react/Library/components/specs/QuickLabelHeader.spec.tsx]: I repectfully disagree - this component is really just sticking together a few divs, it has no logic, the only thing the test does is make sure it doesn't unexpectedly 'look different', which a snapshot can do.

File: app/api/tasks/specs/routes.spec.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        body: { a: 1 },
+      });
+      expect(response.startTime).not.toBe(undefined);
+      response = await routes.get('/api/tasks', { query: { name: 'a' } });
+      while (response.status === 'running') {
+        response = await routes.get('/api/tasks', { query: { name: 'a' } }); // eslint-disable-line
+      }
+      expect(response.message).toBe('1');
+      expect(response.result).toEqual({ a: 1 });
+    });
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/api/tasks/specs/routes.spec.ts]: This retry-until-finished pattern only appears in the test, so I don't think it's worth the extra complexity. In the react app, there is a data refetch every 10 seconds, not just for the state, but also it needs to rerun the elastic search to update the suggestion counts.

File: app/react/Library/components/specs/QuickLabelHeader.spec.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (daneryl) [reply app/react/Library/components/specs/QuickLabelHeader.spec.tsx]: i understand and agree on what snapshots are there for.
The reason i think it is not useful, at least for me is that we have been doing this kind of tests for long time now, never proven to me of utility, only failing whenever im changing the layout on the component on purpose, i see that as implementation details. and sadly building this "muscle memory" of pressing "u" anytime a snapshot fails.

im ok merging this anyways just feeling that snapshot testing has been not proven useful for me.

**Comment by daneryl** [comment]
LGTM

**Comment by gabriel-piles** [comment]
Hi, there is an issue to be fixed. In the quick labeling section when many entities are selected and one of them doesn't have the dictionary (thesauri) to change, the change is possible and a validation error appears

**Comment by bdittes** [comment]
> Hi, there is an issue to be fixed. In the quick labeling section when many entities are selected and one of them doesn't have the dictionary (thesauri) to change, the change is possible and a validation error appears

The validation error occurs when trying to save multilpe values into a (single-)select field. Changing a selection where some documents don't have the thesaurus is possible.

**Pull request merged**